frappe/erpnext
# feat: Enable discount accounting

## Problem:

Currently any item or addition discount applied is accounted in the Income account itself and there is no way to account for discounts in a separate account head. The only workaround possible for now is to use Item Tax templates with negative rate but there is no way to apply fixed amount discount (has to be done manually by back calculating rate) and add that rate in Item Tax Template

## Fix:

- Add a checkbox to _Enable Discount Accounting_ in _Accounts Settings_.
 
![Screenshot 2021-07-06 at 8 27 23 PM](https://user-images.githubusercontent.com/25903035/124622324-a2a43d80-de98-11eb-91b5-8eb397608a19.png)

- If the aforementioned box is checked, the _Discount Account_ field in the _Items_ table for _Sales and Purchase Invoices_ will be made visible using a _Property Setter_. 

![Screenshot 2021-07-06 at 8 31 28 PM](https://user-images.githubusercontent.com/25903035/124623007-36760980-de99-11eb-8d92-ff442fd27463.png)

- This will be allow you to create _GL Entries_ in the following manner:

<details>
<summary>For Sales Invoice</summary>

![Screenshot 2021-07-06 at 8 42 58 PM](https://user-images.githubusercontent.com/25903035/124624722-c49ebf80-de9a-11eb-9791-0b2ce8cd8d07.png)

</details>
<details>
<summary>For Purchase Invoice</summary>

![Screenshot 2021-07-13 at 3 04 48 AM](https://user-images.githubusercontent.com/25903035/125358921-2f626600-e387-11eb-954c-ac7ffdd9e863.png)

</details>

- Additionally, you could enter a _Default Discount Account_ for an _Item_, which will be fetched automatically while creating the Invoice. The visibility for this field is also dependant on the _Enable Discount Accounting_ checkbox in _Accounts Settings_.

![Screenshot 2021-07-06 at 8 46 32 PM](https://user-images.githubusercontent.com/25903035/124625290-4e4e8d00-de9b-11eb-8e46-d5c7a80e9114.png)

- If _Enable Discount Account_ is checked, the _Additional Discount Account_ field will also be made visible. On submitting the invoice, GL Entries will be made against the _Additional Discount Account_ for the _Additional Discount Amount_, as show in the given example.

![Screenshot 2021-07-15 at 9 59 01 PM](https://user-images.githubusercontent.com/25903035/125823867-8e535305-e10b-4c03-b29b-c66e22b7e52d.png)

<details>
<summary>Example</summary>

Item cost = Rs 100
Tax Amount = Rs 20
Discount Percentage = 20%
Therefore, Discount Amount = Rs 24

<details>
<summary>For Sales Invoice</summary>

_Before:_

![Screenshot 2021-07-20 at 2 20 26 AM](https://user-images.githubusercontent.com/25903035/126233217-69bbc644-3611-4dad-a9a7-30a563769048.png)

_After:_

![Screenshot 2021-07-20 at 2 20 04 AM](https://user-images.githubusercontent.com/25903035/126233199-9e2df269-6c8e-426c-b9e9-9c69383e4978.png)

</details>

<details>
<summary>For Purchase Invoice</summary>

_Before:_

![Screenshot 2021-07-20 at 3 31 28 AM](https://user-images.githubusercontent.com/25903035/126233677-ec668e2d-9bdc-4b35-bbc6-813bb462a7fa.png)

_After:_

![Screenshot 2021-07-20 at 3 30 25 AM](https://user-images.githubusercontent.com/25903035/126233724-5357736e-1130-42dd-98bb-1761230218ca.png)

</details>
</details>


<details>
<summary>Docs </summary>

- https://github.com/frappe/erpnext_documentation/pull/361
- https://github.com/frappe/erpnext_documentation/pull/360
 
</details>

<details>
<summary>Testing Info</summary>

1. Open Accounts Settings.
2. Check the Enable Discount Accounting box.
3. Create a new Item/open an existing Item and go to the Sales Defaults section.
4. Enter a Default Discount Account.
5. Create a Sales Invoice for that Item.
6. Enter Discount Amount in the Items table.
7. The Discount Account will be the Item's Default Discount Account, but this can be changed.
8. Save and Submit.
9. Check the GL Entries created and see if they're similar to ones in the image provided above.

</details>
